segmentstore/server/src/main/java/io/pravega/segmentstore/server/writer/AttributeAggregator.java

```diff
@@ -0,0 +1,477 @@
+        this.traceObjectId = String.format("AttributeAggregator[%d-%d]", this.metadata.getContainerId(), this.metadata.getId());
+        this.state = new State(segmentMetadata.getAttributes().getOrDefault(Attributes.ATTRIBUTE_SEGMENT_PERSIST_SEQ_NO, Operation.NO_SEQUENCE_NUMBER));
+        this.closed = new AtomicBoolean();
+        this.lastRootPointer = new AtomicReference<>();
+    }
+
+    //endregion
+
+    //region AutoCloseable Implementation
+
+    @Override
+    public void close() {
+        this.closed.set(true);
+    }
+
+    //endregion
+
+    //region WriterSegmentProcessor Implementation
+
+    @Override
+    public long getLowestUncommittedSequenceNumber() {
+        if (this.lastRootPointer.get() == null) {
+            // There is no async pending update for the root pointer attribute. The LUSN is whatever we accumulated in
+            // our buffers (if nothing, then this will return Operation.NO_SEQUENCE_NUMBER).
+            return this.state.getFirstSequenceNumber();
+        } else {
+            // There is an async pending update for the root pointer attribute. The LUSN can be calculated based off
+            // whatever we were last able to acknowledge.
+            long lpsn = this.state.getLastPersistedSequenceNumber();
+            return lpsn == Operation.NO_SEQUENCE_NUMBER ? this.state.getFirstSequenceNumber() : lpsn + 1;
+        }
+    }
+
+    @Override
+    public boolean isClosed() {
+        return this.closed.get();
+    }
+
+    @Override
+    public String toString() {
+        return String.format("[%d: %s] Count = %d, LUSN = %d, LastSeqNo = %d, LastFlush = %ds", this.metadata.getId(), this.metadata.getName(),
+                this.state.size(), getLowestUncommittedSequenceNumber(), this.state.getLastSequenceNumber(), getElapsedSinceLastFlush().toMillis() / 1000);
+    }
+
+    /**
+     * Adds the given SegmentOperation to the Aggregator.
+     *
+     * @param operation the Operation to add.
+     * @throws DataCorruptionException  If the validation of the given Operation indicates a possible data corruption in
+     *                                  the code (offset gaps, out-of-order operations, etc.)
+     * @throws IllegalArgumentException If the validation of the given Operation indicates a possible non-corrupting bug
+     *                                  in the code.
+     */
+    @Override
+    public void add(SegmentOperation operation) throws DataCorruptionException {
+        Exceptions.checkNotClosed(isClosed(), this);
+        Preconditions.checkArgument(
+                operation.getStreamSegmentId() == this.metadata.getId(),
+                "Operation '%s' refers to a different Segment than this one (%s).", operation, this.metadata.getId());
+        if (isSegmentDeleted()) {
+            return;
+        }
+
+        boolean processed = false;
+        if (operation instanceof StreamSegmentSealOperation) {
```

Comment:
Would an enum be better here?

Reply:
That means I'd need to maintain an enum for the types of Operations, where each operation would have a getter that returns its enum value. That would resolve the `instanceof` but I would still need to cast them eventually to the correct type, so I am not sure we're going to get much benefit out of it. The downside of that would be the need to ensure that the enum matches the operation type; at least with `instanceof` and cast we don't need to worry about that problem.